lib/internal/async_hooks.js

```diff
@@ -357,6 +360,8 @@ function emitAfterScript(asyncId) {
   if (async_hook_fields[kAfter] > 0)
     emitAfterNative(asyncId);
 
+  setCurrentResource(null);
```

Comment:
If we have nested `emitBefore(…, res1)` – `emitBefore(…, res2)` – `emitAfter(…)` – `emitAfter(…)`, then after the first `emitAfter` we want the current resource to be `res1`

Reply:
yeah, shouldn't `setCurrentResource` be `pushCurrentResource`/`popCurrentResource`?

Reply:
> If we have nested emitBefore(…, res1) – emitBefore(…, res2) – emitAfter(…) – emitAfter(…), then after the first emitAfter we want the current resource to be res1

I thought the same. However I struggled to execute user code after `emitAfter`. Can you add an example?

Reply:
I think this should work:
```
const async_hooks = require("async_hooks");
const fs = require("fs");

let indent = 0;

function init(asyncId, type, triggerAsyncId, resource) {
  const eid = async_hooks.executionAsyncId();
  const indentStr = ' '.repeat(indent);
  fs.writeSync(1, `${indentStr}${type}(${asyncId}): trigger: ${triggerAsyncId} execution: ${eid}\n`);
}

function before(asyncId) {
  const indentStr = ' '.repeat(indent);
  fs.writeSync(1, `${indentStr}before:  ${asyncId}\n`);
  indent += 2;
}

function after(asyncId) {
  indent -= 2;
  const indentStr = ' '.repeat(indent);
  fs.writeSync(1, `${indentStr}after:   ${asyncId}\n`);
}

function destroy(asyncId) {
  const indentStr = ' '.repeat(indent);
  fs.writeSync(1, `${indentStr}destroy: ${asyncId}\n`);
}

async_hooks.createHook({ init, before, after, destroy }).enable();

class MyResource extends async_hooks.AsyncResource {
  constructor() {
    super("XXX");
  }
}
const res = new MyResource();

setImmediate(() => {
  const indentStr = ' '.repeat(indent);
  fs.writeSync(1, `${indentStr}between before\n`);
  res.runInAsyncScope(() => {
    const indentStr = ' '.repeat(indent);
    fs.writeSync(1, `${indentStr}inner \n`);
  });
  fs.writeSync(1, `${indentStr}between after\n`);
});
```

Reply:
Is this only happening with runInAsyncScope? That would be easy to fix without adding a full stack where we push/pop.

Adding a full stack is also a serious refactoring of the async_hooks machinery, something that this PR currently avoids.

Reply:
@mcollina We do have a full stack in place currently, though?

> Is this only happening with runInAsyncScope?

It can happen with all mechanisms that trigger async hooks.